File: Atcoder/ABC393/f.cpp
#include <bits/stdc++.h>
using namespace std;
typedef long long ll;
typedef tuple<ll, ll, ll> tp;
typedef pair<ll, ll> pll;

int main(void)
{
	cin.tie(0)->sync_with_stdio(0);
	ll n = 0, q = 0;
	cin >> n >> q;
	vector<pll>v;
	for (ll i = 1; i <= n; ++i)
	{
		ll temp = 0;
		cin >> temp;
		v.push_back(make_pair(temp, i));
	}
	vector<tp>Q(q);
	for (ll i = 0; i < q; ++i)
	{
		auto& [a, b, c] = Q[i];
		cin >> a >> b;
		c = i;
	}
	sort(v.begin(), v.end(), [&](const pll& a, const pll& b)
		{
			return a.first < b.first;
		});
	sort(Q.begin(), Q.end(), [&](const tp& a, const tp& b)
		{
			return get<1>(a) < get<1>(b);
		});
	vector<ll>BIT(n + 1);
	auto update = [&](ll idx, ll delta)
		{
			while (idx <= n)
			{
				BIT[idx] = max(BIT[idx], delta);
				idx += idx & -idx;
			}
		};
	auto query = [&](ll idx)
		{
			ll res = 0;
			while (idx > 0)
			{
				res = max(res, BIT[idx]);
				idx -= idx & -idx;
			}
			return res;
		};
	vector<ll>ans(q + 1, 0);
	ll jdx = 0, idx = 0;
	while (idx < q)
	{
		auto& [a, b, c] = Q[idx];
		while (jdx < n && v[jdx].first <= b)
		{
			vector<ll>temp;
			ll pivot = v[jdx].first;
			while (jdx < n && v[jdx].first == pivot)
			{
				temp.push_back(v[jdx].second);
				jdx++;
			}
			vector<ll>p;
			for (auto& i : temp)
			{
				p.push_back(query(i - 1) + 1);
			}
			for (ll i = 0; i < (ll)p.size(); ++i)
			{
				update(temp[i], p[i]);
			}
		}
		ans[c] = query(a);
		idx++;
	}
	for (ll i = 0; i < q; ++i)
	{
		cout << ans[i] << '\n';
	}
	return 0;
}

File: Atcoder/ABC396/f.py
import sys
from collections import defaultdict
from collections import deque
input = sys.stdin.readline
n,m = map(int,input().strip().split())
v = list(map(int,input().strip().split()))
BIT = [0 for _ in range(m+1)]
ans = [0 for _ in range(m)]
temp = [[] for _ in range(m)]
for i in range(n):
    temp[v[i]].append(i+1)
def update (idx, delta):
    while idx <= m:
        BIT[idx] += delta
        idx += idx & -idx
def query (idx):
    res = 0
    while idx:
        res += BIT[idx]
        idx -= idx & -idx
    return res
for i in v:
    ans[0] += query(m) - query(i+1)
    update(i+1,1)
now = ans[0]
for i in range(m - 1):
    pivot = 0
    for j in temp[m - i - 1]:
        pivot += 2 * j - n - 1
    now += pivot
    ans[i+1] = now
for i in ans:
    print(i)
